### PR TITLE
Feature: add disk quota gc for daemon

### PR DESCRIPTION
**Pull request by jim3ma**
Signed-off-by: Jim Ma <majinjing3@gmail.com>

## Related Issue

#214 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [x] Code compiles correctly.
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by gaius-qi — DISMISSED** [review]

File: client/daemon/storage/storage_manager.go [hunk 1]
```diff
@@ -465,6 +469,30 @@ func (s *storageManager) TryGC() (bool, error) {
 		}
 		return true
 	})
+
+	if s.storeOption.DiskGCThreshold.SizeInBytes > 0 && totalSize > s.storeOption.DiskGCThreshold.SizeInBytes {
```
Comment (244372610) [inline client/daemon/storage/storage_manager.go:473]: 这里是否需要先回收完过期的任务后再去判断空间是否不足，现在这种是否会存在过多的删除任务

File: client/daemon/storage/storage_manager.go [hunk 2]
```diff
@@ -465,6 +469,30 @@ func (s *storageManager) TryGC() (bool, error) {
...
+				task.TaskID, task.PeerID, task.lastAccess.Format(time.RFC3339Nano),
+				units.BytesSize(float64(task.ContentLength)))
+			totalSize -= task.ContentLength
+			if totalSize < s.storeOption.DiskGCThreshold.SizeInBytes {
+				break
+			}
+		}
+	}
+
 	for _, key := range s.markedReclaimTasks {
```
Comment (244372610) [inline client/daemon/storage/storage_manager.go:496]: 执行回收的时候，是否应该先判断当前的可回收状态是否发生了变更？防止在前面判断可以回收了，但是在真正回收的之前状态又发生了变化

Comment (jim3ma) [reply client/daemon/storage/storage_manager.go:496]: 这里是回收上一轮周期 mark 的 task，不是这一轮的

File: client/daemon/storage/storage_manager.go (hunk 1 above)
Comment (jim3ma) [reply client/daemon/storage/storage_manager.go:473]: 嗯，在 task expire time 检查了就要过滤掉这些了

File: client/daemon/storage/storage_manager.go (hunk 2 above)
Comment (244372610) [reply client/daemon/storage/storage_manager.go:496]: 但在执行回收前还是要判断下当前的状态吧，不然在你插入到可回收列表后到真正执行的这段时间可能该任务又被访问了

Comment (jim3ma) [reply client/daemon/storage/storage_manager.go:500]: leavetask 之后，如果在等待回收的时候，还有访问，调度器会重新调取的

**Review by gaius-qi — APPROVED** [review]

**Review by 244372610 — APPROVED** [review]

**Pull request merged**